[processor/k8sattributes] assume current container if restart_count i…

ty-elastic opened the pull request:
Assume current instance of container when looking up `container.id` if `k8s.container.restart_count` isn't explicitly specified.

**Description:** <Describe what has changed.>
* check if `k8s.container.restart_count` is asserted
* if it is, only match container to specified restart_count
* if it is not, match container to currently running container

**Link to tracking Issue:** #16431

**Testing:**
* asserted `k8s.container.restart_count=0` and verified that code still matches only the first instance of the container
* did not assert `k8s.container.restart_count` and verified that code matches the currently running instance of container

**Documentation:**
* where would be the most helpful place to explain this behavior?

ty-elastic committed:
[processor/k8sattributes] assume current container if restart_count isn't specified

ty-elastic commented:
@dashpole, @dmitryax , any thoughts? see https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16431 for rationale.

ricramalho commented:
Hi. My team is really needing this feature to be merged. We are making a OpenTelemetry integration with Datadog to collect logs, metrics and traces, and we would appreciate to have the container id attribute to get the best out of all the dashboards.
We have OpenTelemetry collector running as a sidecar on the pods we want to log to Datadog - we configured the shared volume mounts to get the logs, receiver and k8sprocessor to get more details regarding the pod/containers. If we add the container name and restart count as resource attributes, it works until we restart the container:

resource:
        attributes:
        - key: k8s.container.restart_count #mandatory
          value: 0
          action: insert
        - key: k8s.container.name #mandatory
          value: dummy-nginx
          action: insert
 
Once we restart the container, it stops sending the container id. If we manage to get the latest container ID (with the higher restart count), that would solve our problem.
Right now we are testing this with the restart count set to 0, however, as you know, it's not feasible to have it this way using in higher environments. Thank you.

dashpole commented:
cc codeowners @dmitryax @rmfitzpatrick

ty-elastic commented:
On my end, I'll rebase the PR and add changelog and tests.

dmitryax commented on processor/k8sattributesprocessor/processor.go:
```diff
@@ -167,17 +167,28 @@ func (kp *kubernetesprocessor) addContainerAttributes(attrs pcommon.Map, pod *ku
+			runID = containerRunID
+		}
+	} else {
+		// find runID (restart count) for currently running container
```
```suggestion
		// take the highest runID (restart count) which represents the currently running container in most cases
```

ty-elastic replied:
thx, @dmitryax! added changelog entry and test coverage, awaiting tests to pass...

dmitryax reviewed (commented):
@ty-elastic please add a changelog and test coverage, then we can merge it

ty-elastic commented:
bump. changelog and test cases updated. awaiting review.

mstraubinger-deloitte commented:
bump. This merge would be highly appreciated.

ty-elastic commented:
cc codeowners @owais  @rmfitzpatrick

dmitryax commented on CHANGELOG.md:
```diff
@@ -2,7 +2,9 @@
-<!-- next version -->
+### 💡 Enhancements 💡
+
+- `k8sattributesprocessor`: Default `container.id` to latest container instance unless `k8s.container.restart_count` is specified. (#16432)
```
Sorry for the delay. Please add changelog entry to https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/.chloggen 

This file is auto-generated

dmitryax commented:
Given that `container.id` is enabled by default, I'm worried that we start emitting it for users that don't send `container.restart_count` and don't expect that attribute to be added. Since we cannot assure it's accurate all the time now, I suggest removing `container.id` from the default `metadata` list, please create a `breaking_change` changelog entry specifically for that.

Also, please update the documentation in https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/k8sattributesprocessor/doc.go#L70-L72

ty-elastic commented:
Hi @dmitryax , I appreciate the comments and feedback. I've updated the changelog (properly), and per your suggestion, removed `container.id` as default metadata (flagged as a breaking change).

dmitryax reviewed (commented):
Please remove [this line](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/k8sattributesprocessor/config.go#L91) as well. Otherwise, LGTM

ty-elastic commented:
> Please remove [this line](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/k8sattributesprocessor/config.go#L91) as well. Otherwise, LGTM

my apologies; this has been addressed. thank you!!
